File: 2015/1/pa.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
using namespace std;

int main() {
    ifstream fin("input.txt");
    string line;
    getline(fin, line);
    vector<char> dir(line.begin(), line.end());

    int count = 0;
    for (auto x : dir) {
        if (x == '(') count++;
        if (x == ')') count--;
    }

    cout << count << endl;

    return 0;
}

File: 2016-py/day4/pa.py
from collections import Counter
import re


def main():
    total = 0

    with open("input.txt") as f:
        for line in f:
            line = line.strip()

            # Extract name, sector ID, and checksum
            m = re.match(r"([a-z-]+)-(\d+)\[([a-z]+)\]", line)

            name = m.group(1)
            sector = int(m.group(2))
            checksum = m.group(3)

            # Count letters only
            freq = Counter(c for c in name if c.isalpha())

            # Sort by frequency, then alphabetically
            ordered = sorted(freq.items(), key=lambda x: (-x[1], x[0]))

            # Build the computed checksum
            computed = "".join(char for char, _ in ordered[:5])

            if computed == checksum:
                total += sector

    print(total)


if __name__ == "__main__":
    main()

